engine/src/main/java/org/terasology/engine/core/module/ModuleManager.java

```diff
@@ -105,49 +98,61 @@ public ModuleManager(String masterServerAddress, List<Class<?>> classesOnClasspa
 
     /**
      * I wondered why this is important, and found MovingBlocks/Terasology#1450.
-     * It's not a worry that the engine module wouldn't be loaded without it. 
-     * It's about ordering: some things run in an order derived from the dependency 
+     * It's not a worry that the engine module wouldn't be loaded without it.
+     * It's about ordering: some things run in an order derived from the dependency
      * tree, and we want to make sure engine is at the root of it.
-     */   
-    private void ensureModulesDependOnEngine(Module engineModule) {
+     */
+    private void ensureModulesDependOnEngine() {
         DependencyInfo engineDep = new DependencyInfo();
         engineDep.setId(engineModule.getId());
         engineDep.setMinVersion(engineModule.getVersion());
-        engineDep.setMaxVersion(engineModule.getVersion().getNextPatchVersion());
+        engineDep.setMaxVersion(engineModule.getVersion().getNextMinorVersion());
+
+        HashSet<Name> engineModules = Sets.newHashSet(engineModule.getId());
+        engineModules.addAll(engineModule.getMetadata().getDependencies().stream().map(DependencyInfo::getId).collect(Collectors.toList()));
 
-        registry.stream().filter(mod -> mod != engineModule).forEach(mod -> mod.getMetadata().getDependencies().add(engineDep));
+        registry.stream()
+                .filter(mod -> !engineModules.contains(mod.getId()))
+                .forEach(mod -> mod.getMetadata().getDependencies().add(engineDep));
     }
 
     private void loadModulesFromApplicationPath(PathManager pathManager) {
-        ModulePathScanner scanner = new ModulePathScanner(new ModuleLoader(metadataReader));
-        scanner.getModuleLoader().setModuleInfoPath(TerasologyConstants.MODULE_INFO_FILENAME);
-        scanner.scan(registry, pathManager.getModulePaths());
+        ModulePathScanner scanner = new ModulePathScanner(moduleFactory);
+        List<File> paths = pathManager.getModulePaths().stream().map(Path::toFile).collect(Collectors.toList());
+        scanner.scan(registry, paths);
     }
 
     private Module loadEngineModule(List<Class<?>> classesOnClasspathsToAddToEngine) {
-        Module engineModule;
-        try (Reader reader = new InputStreamReader(getClass().getResourceAsStream("/engine-module.txt"), TerasologyConstants.CHARSET)) {
-            ModuleMetadata metadata = metadataReader.read(reader);
-            List<Class<?>> additionalClassesList = new ArrayList<>(classesOnClasspathsToAddToEngine.size() + 2);
-            additionalClassesList.add(Module.class); // provide access to gestalt-module.jar
-            additionalClassesList.add(Asset.class); // provide access to gestalt-asset-core.jar
-            additionalClassesList.add(UIWidget.class); // provide access to nui.jar
```

Comment:
So how does/did this work? You just need to provide a single entry-point class, and the rest is collected automatically? (Guess I'll need a lecture on reflections and class loading...)

Reply:
These go through reflections.ClasspathHelper.forClass that returns a URL that "contains the class."

That is often a `.jar` that contains that `.class`, so you pass `UIWidget.class` and then Reflections will do its scanning thing on everything in `nui.jar`.

but I definitely have concerns about how well this works in all cases.

What if you're running under IntelliJ and it has put `classes/` directories on the classpath instead of jars?

What if you're running from a distribution build with `jpackage` and it's bundled everything in to one archive or fat jar or something?

The Reflections stuff and the APIs we've built on top of them poke through the abstraction layer that usually keeps us from worrying too much about how the class files are organized, and I am uneasy about it.

Reply:
Okay, so there are a bunch of assumptions that must hold for this to work (and typically it will work in our workspaces and in distributions).

The classes we mention here, however, are handpicked "representatives" of each of the artifacts we want to allow-list. If we (re-)move that class, things will break, but at least it will break at compilation time (actual class references here, not just string identifiers or such)...

Well, at least there are still people around who understand this :upside_down_face: